File: testdata/fixtures/javascript/vulnerable/log_injection.ts
import { Request, Response } from 'express';
import winston from 'winston';

const logger = winston.createLogger({
  level: 'info',
  transports: [new winston.transports.Console()],
});

// VULNERABLE: User input logged directly without sanitization
export function loginAttempt(req: Request, res: Response): void {
  const { username } = req.body;

  // VULNERABLE: console.log with unsanitized request data (log injection)
  // Attacker can inject: "admin\n[INFO] Login successful for admin"
  console.log('Login attempt for user: ' + req.body.username);

  // VULNERABLE: winston logger with unsanitized request data
  logger.info(`Failed login attempt for user: ${username} from IP: ${req.headers['x-forwarded-for']}`);

  res.status(401).json({ error: 'Invalid credentials' });
}

// VULNERABLE: Logging sensitive data (password)
export function debugLogin(req: Request, res: Response): void {
  const { username, password } = req.body;

  // VULNERABLE: Sensitive data in logs - logging password
  console.log('Debug auth - username: ' + username + ', password: ' + password);
  logger.debug(`Authentication attempt with credentials - password: ${password}`);

  res.json({ status: 'ok' });
}

File: testdata/fixtures/go/vulnerable/log_injection.go
package vulnerable

import (
	"log"
	"net/http"
)

// VULN: Log injection - unsanitized user input written directly to log.
// Should trigger GTSS-LOG-001 (Unsanitized User Input in Log Calls).

func HandleLogin(w http.ResponseWriter, r *http.Request) {
	username := r.FormValue("username")
	password := r.FormValue("password")

	// Attacker can inject newlines: username = "admin\n[INFO] Login successful for admin"
	log.Printf("Login attempt for user: %s from IP: %s", username, r.RemoteAddr)

	if !authenticate(username, password) {
		log.Printf("Failed login for user: %s", r.FormValue("username"))
		http.Error(w, "invalid credentials", http.StatusUnauthorized)
		return
	}

	log.Printf("Successful login for user: %s", username)
	w.Write([]byte("welcome"))
}

func authenticate(user, pass string) bool {
	return user == "admin" && pass == "admin"
}
